radix-engine-common/src/types/type_identifier.rs

```diff
@@ -5,15 +5,39 @@ define_wrapped_hash!(
     SchemaHash
 );
 
-/// An Identifier for a structural type. This can be treated almost
-/// like a pointer as two equivalent type identifiers will map to
-/// the same schema
+/// A global identifier for a type in a Radix network.
+/// A type is either well-known, or local to a node.
+/// This identifier includes the NodeId, which provides context for how
+/// to look-up the type.
+///
+/// If/when we add additional type metadata (eg translations, documentation),
+/// these will be added by the owner of the Node against the GlobalTypeAddress.
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Sbor)]
-pub struct TypeIdentifier(pub SchemaHash, pub LocalTypeIndex);
+pub enum GlobalTypeAddress {
+    WellKnown(WellKnownTypeId),
+    NodeLocal(NodeId, SchemaHash, usize),
+}
+
+/// An identifier for a type under a given schema context in the Radix network.
+///
+/// Note - this type provides scoping to a schema even for well-known types where
+/// the schema is irrelevant.
+///
+/// See also:
+/// * [`GlobalTypeAddress`] which captures the global type identity of a type,
+/// for the purpose of semantics and ownership.
+/// * [`NodeScopedTypeId`] which captures an identifier for a type where the node
+/// is clear from context.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Sbor)]
+pub struct FullyScopedTypeId(pub NodeId, pub SchemaHash, pub LocalTypeId);
+
+/// An identifier for a type in the context of a schema under a given node.
+/// The given node is context-dependent.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Sbor)]
+pub struct NodeScopedTypeId(pub SchemaHash, pub LocalTypeId);
```

Comment:
I dislike `TypeIdentifier` because we have so many type identifiers, and it's not clear what it is.

If you want `SystemTypeId` or something, we can. But tbh the NodeId / Full address is important.

Reply:
hrm true, how about `HashTypeId`

Reply:
Have settled for `ScopedTypeId` after discussion